src/main/java/network/brightspots/rcv/ClearBallotCvrReader.java

```
@@ -0,0 +1,144 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+ * the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with this
+ * program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package network.brightspots.rcv;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import javafx.util.Pair;
+
+import network.brightspots.rcv.CastVoteRecord.CvrParseException;
+
+
+class ClearBallotCvrReader {
+
+  private final String cvrPath;
+  private final ContestConfig contestConfig;
+
+  ClearBallotCvrReader(String cvrPath, ContestConfig contestConfig) {
+    this.cvrPath = cvrPath;
+    this.contestConfig = contestConfig;
+  }
+
+  // parse Cvr json into CastVoteRecord objects and append them to the input castVoteRecords list
+  // see Clear Ballot 2.1 RCV Format Specification for details
+void readCastVoteRecords(List<CastVoteRecord> castVoteRecords) throws CvrParseException {
+    BufferedReader csvReader;
+    try {
+      csvReader = new BufferedReader(new FileReader(this.cvrPath));
+      // each "choice column" in the input Csv corresponds to a particular ranking: candidate+rank pair
+      // we parse these rankings from the header row into a map for lookup during CVR parsing
+      String firstRow = csvReader.readLine();
+      if (firstRow == null) {
+        Logger.log(Level.SEVERE, "No header row found in cast vote record file: %s", this.cvrPath);
+        throw new CvrParseException();
+      }
+      String[] headerData = firstRow.split(",");
+      if (headerData.length < CvrColumnField.ChoicesBegin.ordinal()) {
+        Logger.log(Level.SEVERE, "No choice columns found in cast vote record file: %s", this.cvrPath);
```

should we throw or otherwise return an error here instead of just logging?